### PR TITLE
SWATCH-1237: Use swatch-product-configuration in subscription syncing

### PR DESCRIPTION
**Pull request by kahowell**
Jira issue: [SWATCH-1237](https://issues.redhat.com/browse/SWATCH-1237)

Description
===========

Subscription syncing used tag profile for the following things:

* Log a warning if a PAYG sub is missing details
* Log a warning if attempting to terminate a PAYG sub more than an hour out
* Lookup the product names used for a PAYG sub lookup
* Provide a lookup for productIds/tags for a SKU.
* Provide a lookup for metrics for a productId/tag.

Testing
=======

1. Deploy the service:

```shell
DEV_MODE=true \
  DEVTEST_EVENT_EDITING_ENABLED=true \
  SUBSCRIPTION_USE_STUB=true \
  PRODUCT_USE_STUB=true \
  ./gradlew :bootRun
```

2. Sync subscriptions:

```shell
http PUT :8000/api/rhsm-subscriptions/v1/internal/subscriptions/sync/org/123 \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

3. Terminate a subscription:

```shell
http POST :8000/api/rhsm-subscriptions/v1/internal/subscriptions/terminate/235251 \
  timestamp==2030-01-01T00:00Z \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

4. Lookup aws usage context:

```shell
http :8000/api/rhsm-subscriptions/v1/internal/subscriptions/awsUsageContext \
  orgId==123 \
  productId==rosa \
  date==2023-01-01T00:00Z \
  awsAccountId==foo \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

See in the logs a line like:
> No subscription found for account null with criteria DbReportCriteria(accountNumber=null, orgId=123, productTag=null, productNames=[OpenShift Dedicated], productId=null, serviceLevel=EMPTY, usage=_ANY, billingProvider=AWS, billingAccountId=foo, payg=true, beginning=2022-12-31T23:00Z, ending=2023-01-01T00:00Z, metricId=null, hypervisorReportCategory=null)

(Verify that `OpenShift Dedicated` is in `productNames`)

5. Lookup tags for a SKU:

```shell
http :8000/api/rhsm-subscriptions/v1/internal/offerings/MW01485/product_tags \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

(See `OpenShift-metrics` in the response).

6. Lookup metrics for a tag:

```shell
http :8000/api/rhsm-subscriptions/v1/internal/tags/rosa/metrics \
  x-rh-swatch-psk:placeholder \
  Origin:console.redhat.com
```

(Verify `Instance-hours` and `Cores` in the response).

Verification
------------

* Verify warning log when PAYG sub is missing details, message like:
> PAYG eligible subscription with subscriptionId:235251 has no billing provider.

* Verify warning log when attempting to terminate a PAYG sub more than an hour out, message like:
> Subscription 235251 terminated at 2023-08-17T19:50:06.974839779Z with out of range termination date 2030-01-01T00:00Z

Verify api calls succeed w/ non-empty results for lookups.

### CODE REVIEW

**Review by ntkathole — APPROVED** [review]
Looks good!

Subscription related internal rest calls are working fine, will need https://github.com/RedHatInsights/rhsm-subscriptions/pull/2406 to verify actual numbers, will do further testing in stage or once that pr merged.

File: src/main/java/org/candlepin/subscriptions/capacity/admin/TagMetricMapper.java [hunk 1]
```diff
@@ -20,13 +20,17 @@
...
 
 @Mapper(componentModel = "spring")
 public interface TagMetricMapper {
 
-  List<org.candlepin.subscriptions.utilization.admin.api.model.TagMetric> mapTagMetrics(
-      List<TagMetric> metricsForTag);
+  List<TagMetric> mapTagMetrics(List<Metric> metricsForTag);
+
+  @Mapping(target = "uom", source = "id")
+  TagMetric fromConfigurationMetric(Metric metric);
```
Comment (lindseyburnett) [inline src/main/java/org/candlepin/subscriptions/capacity/admin/TagMetricMapper.java:35]: Did this get used anywhere?

File: src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java [hunk 2]
```diff
@@ -706,9 +710,11 @@ public OfferingProductTags findProductTags(String sku) {
     return productTags;
   }
 
-  public List<TagMetric> getMetricsForTag(String tag) {
-    return tagProfile.getTagMetrics().stream()
-        .filter(tagMetric -> Objects.equals(tagMetric.getTag(), tag))
+  public List<Metric> getMetricsForTag(String tag) {
```
Comment (kartikshahc) [inline src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java:713]: Feels like helper methods that would fit in Variant class?

**Review by kartikshahc — CHANGES_REQUESTED** [review]

File: src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java [hunk 3]
```diff
@@ -628,7 +627,11 @@ public List<org.candlepin.subscriptions.db.model.Subscription> findSubscriptions
     Assert.isTrue(ServiceLevel._ANY != usageKey.getSla(), "Service Level cannot be _ANY");
 
     String productId = usageKey.getProductId();
-    Set<String> productNames = tagProfile.getOfferingProductNamesForTag(productId);
+    Set<String> productNames =
+        Variant.findByTag(productId).stream()
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java:631]: Will add to Variant as `getProductNamesForTag`

File: src/main/java/org/candlepin/subscriptions/capacity/admin/TagMetricMapper.java [hunk 4]
```diff
@@ -20,13 +20,17 @@
...
+import org.candlepin.subscriptions.utilization.admin.api.model.TagMetric;
 import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
 
 @Mapper(componentModel = "spring")
 public interface TagMetricMapper {
 
-  List<org.candlepin.subscriptions.utilization.admin.api.model.TagMetric> mapTagMetrics(
-      List<TagMetric> metricsForTag);
+  List<TagMetric> mapTagMetrics(List<Metric> metricsForTag);
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/capacity/admin/TagMetricMapper.java:32]: :thinking: yeah, will add another commit with changes for this.

File: src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java [hunk 5]
```diff
@@ -256,9 +253,10 @@ private void checkForMissingBillingProvider(
         || subscription.getBillingProvider().equals(BillingProvider.EMPTY)) {
       // The offering here is going to be a proxy object created by getReferenceById.  Hibernate
       // should take care of actually performing the select from the database if one is needed.
-      var productTag =
-          tagProfile.tagForOfferingProductName(subscription.getOffering().getProductName());
-      if (tagProfile.isProductPAYGEligible(productTag)) {
+      var subscriptionDefinition =
+          SubscriptionDefinition.lookupSubscriptionByProductName(
+              subscription.getOffering().getProductName());
+      if (subscriptionDefinition.isPresent() && subscriptionDefinition.get().isPaygEnabled()) {
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java:259]: Nah, it's fine as-is, there will be many cases where there is a product name on the offering but there's no matching configuration.

**Comment by kahowell** [comment]
@kartikshahc please take another look when you get a chance.

**Review by lindseyburnett — APPROVED** [review]

**Review by kartikshahc — APPROVED** [review]
LGTM

**Pull request merged**